File: java/binary-search/src/main/java/BinarySearch.java
import java.util.List;

public class BinarySearch<T extends Comparable<T>> {

    private final List<T> input;

    public BinarySearch(List<T> input) {
        this.input = input;
    }

    public int indexOf(T value) {
        if (input.isEmpty())
            return -1;

        return indexOf(value, 0, input.size() - 1);
    }

    public int indexOf(T value, int minIndex, int maxIndex) {
        int middleIndex = (minIndex + maxIndex) / 2;

        if (input.get(middleIndex).compareTo(value) == 0)
            return middleIndex;

        if (middleIndex <= 0 || middleIndex >= input.size() - 1 || middleIndex < minIndex)
            return -1;

        if (input.get(middleIndex).compareTo(value) > 0)
            return indexOf(value, minIndex, middleIndex - 1);

        return indexOf(value, middleIndex + 1, maxIndex);
    }
}

File: ecmascript/binary-search/binary-search.js
Array.prototype.isSorted = function() {
    for (let i = 0; i < this.length - 2; i++) {
        if (this[i] > this[i + 1])
            return false;
    }

    return true;
}

export default class BinarySearch {
    constructor(input) {
        this._array = input.isSorted() ? input : undefined;
    }

    get array() {
        return this._array;
    }

    indexOf(value) {
        if (this.array.length === 0)
            return -1;

        return this.indexOfRange(value, 0, this.array.length - 1);
    }

    indexOfRange(value, minIndex, maxIndex) {
         var middleIndex = Math.round((minIndex + maxIndex) / 2);

        if (this.array[middleIndex] == value)
            return middleIndex;

        if (middleIndex <= 0 || middleIndex >= this.array.length - 1 || middleIndex < minIndex)
            return -1;

        if (this.array[middleIndex] > value)
            return this.indexOfRange(value, minIndex, middleIndex - 1);

        return this.indexOfRange(value, middleIndex + 1, maxIndex);
    }
}